torch/csrc/jit/script/compiler.cpp

```diff
@@ -686,16 +686,36 @@ struct to_ir {
     // inputs
     auto it = def.params().begin();
     auto end = def.params().end();
+    // Type annotations exclude explicitly typing the "self" parameter, so in the
+    // case that this is a method with self we expect one fewer parameter annotation
+    // than the number of parameters this Def takes.
+    auto expected_annotation_size = self ? def.params().size() - 1 : def.params().size();
+    if (typed_def.schema && typed_def.schema->arguments.size() != expected_annotation_size) {
+      throw ErrorReport(def.params().range()) << "Number of type annotations for"
+        << " function parameters (" << typed_def.schema->arguments.size() << ")"
+        << " does not match the number of parameters on the function ("
+        << expected_annotation_size << ")!";
+    }
     if(self) {
       if(it == end)
         throw ErrorReport(def.params().range()) << "methods must have a self argument";
       environment_stack->setSugaredVar(def.range(), (*it).ident().name(), self);
       ++it;
     }
+    size_t arg_annotation_idx = 0;
     for(;it != end; ++it) {
       auto& name = (*it).ident().name();
-      arguments.push_back({name, DynamicType::get()});
-      environment_stack->setVar((*it).ident().range(), name, graph->addInput(name));
+      // Add the input to the graph
+      Value *new_input = graph->addInput(name);
+      environment_stack->setVar((*it).ident().range(), name, new_input);
+
+      // Record the type for the schema and set the Type on the Value*
+      TypePtr arg_type = DynamicType::get();
+      if (typed_def.schema) {
+        arg_type = typed_def.schema->arguments[arg_annotation_idx++].type;
```

Comment:
nit: can we use `.at` instead of indexing here? Just to make sure we won't break this logic in the future and get UB silently